shared/src/sources/balancer/pool_init.rs

```diff
@@ -27,12 +32,49 @@ pub trait PoolInitializing: Send + Sync {
 ///
 /// This can be used to index all pools from events instead of relying on the
 /// Balancer subgraph for example.
-pub struct EmptyPoolInitializer;
+pub struct EmptyPoolInitializer(u64);
 
 #[async_trait::async_trait]
 impl PoolInitializing for EmptyPoolInitializer {
     async fn initialize_pools(&self) -> Result<BalancerRegisteredPools> {
-        Ok(Default::default())
+        let fetched_block_number = deployment_block(BalancerV2Vault::artifact(), self.0).await?;
+        Ok(BalancerRegisteredPools {
+            fetched_block_number,
+            ..Default::default()
+        })
+    }
+}
+
+/// The default Balancer pool initializer.
+pub enum DefaultPoolInitializer {
+    Subgraph(SubgraphPoolInitializer),
+    Empty(EmptyPoolInitializer),
+}
+
+impl DefaultPoolInitializer {
+    pub fn new(chain_id: u64) -> Result<Self> {
+        const MAINNET_CHAIN_ID: u64 = 1;
+
+        Ok(match chain_id {
+            // Balancer subgraph seems to only correctly index pool info on
+            // chains where it supports archive nodes (because of the required
+            // `eth_call`s). This means we can only use the pure Subgraph
+            // initializer on Mainnet.
+            MAINNET_CHAIN_ID => {
+                DefaultPoolInitializer::Subgraph(SubgraphPoolInitializer::new(chain_id)?)
+            }
```

Comment:
I think this just binds chain_id to the constant and not check for equality which you probably intended.

Reply:
Changed this to use an `if` statement instead so there is less ambiguity.